### ISSUE TITLE
Wrong cursors on Xorg

### ISSUE THREAD

**Issue by kra-mo** [libsdl-org/SDL#8939]
GNOME recently [dropped support for](https://gitlab.gnome.org/GNOME/adwaita-icon-theme/-/merge_requests/61) X11 cursor names. As a result, these cursors fall back to the ugly X11 defaults even though they have alternatives installed on the system.

Cursors from the [Freedesktop cursor spec](https://www.freedesktop.org/wiki/Specifications/cursor-spec/) should be checked for first, [like they are on Wayland](https://github.com/libsdl-org/SDL/blob/9f7eb6c4c11b445da0fcadb357e1c7ed053b0084/src/video/wayland/SDL_waylandmouse.c#L321), only falling back to X11 names if strictly necessary.

**Comment by icculus** [comment]
@slouken, can we try to slip a fix for this into 2.30.0, or are we out of time at this point?

**Comment by slouken** [comment]
Yep, we just need a tested PR

**Comment by vovan4ik123-engine** [comment]
Boyz thanks for your work. We wait release =).

**Comment by icculus** [comment]
So this doesn't use Gnome names at all right now, just the Xlib API from 40 years ago:

https://github.com/libsdl-org/SDL/blob/9f7eb6c4c11b445da0fcadb357e1c7ed053b0084/src/video/x11/SDL_x11mouse.c#L224-L286

So maybe we can improve this, but this isn't something that _regressed_ when Gnome changed, so I'm taking this out of the 2.30.0 milestone.

**Comment by kra-mo** [comment]
I don't understand. GNOME definitely did use to provide cursors with these names and now they don't, leading to broken-looking cursors. What do you mean there is no regression?

**Comment by slouken** [comment]
I'm bumping this back into the release milestone, hopefully fixing this is just falling back to GNOME cursors if the old ones don't exist?

**Comment by slouken** [comment]
Okay, fixing this involves adding a dependency on the XCursor library and using XcursorLibraryLoadCursor(). This is something we definitely don't want at the last minute in 2.30.

**Comment by icculus** [comment]
I'm saying our X11 code gets these cursors from Xlib, not Gnome, using a function call that predates Gnome.

We could maybe add Gnome-specific code, but at the moment that code doesn't exist.

**Comment by kra-mo** [comment]
The code would never be GNOME-specific. This is about the icon theme GNOME uses, which lacks X cursor names now.

**Comment by smcv** [comment]
> this doesn't use Gnome names at all right now, just the Xlib API from 40 years ago

I suspect that what has changed here might be: previously, using named cursors from the Xlib API would have caused GNOME to substitute something nicer-looking and hopefully somewhere close to semantically equivalent, from a modern icon theme. Now, it's no longer doing that, and we get authentically 40 year old Xlib behaviour.

**Comment by smcv** [comment]
The non-default Wayland video backend in SDL2 has essentially the same issue: it's assuming that the cursor theme contains 1980s X11 cursor names, and not the CSS names (which are also what's used in the freedesktop.org cursor spec, I don't know which one was inspired by the other).

The GNOME change here is that older versions of the Adwaita icon theme shipped cursors with both X11 and CSS names, but version 46 only provides the CSS names.

In Wayland, this should at least be an easy fix with no new dependencies.

**Comment by icculus** [comment]
I'm going to take a run at the Xcursor stuff tonight and report back. Hopefully it's simple enough.

**Comment by icculus** [comment]
Okay, I added something that _should_ fix the problem; it can load the correct cursors based on CSS theme names here, using Xcursor, but the Gnome on my system still has the old names available, so I'd appreciate it if someone could test this on a newer Gnome install, just to make sure the problem is definitely resolved.

**Comment by icculus** [comment]
(Pushed to SDL2 and release-2.30.x branches, too.)

**Comment by icculus** [comment]
Oh, Wayland needs fixing, too, reopening.

**Comment by icculus** [comment]
Here's an untested patch for Wayland, for SDL2.

Note that SDL3's Wayland code already uses the CSS names exclusively with no fallback, so I'm not sure if the fallback is unnecessary for older Wayland installs or we're just throwing caution to the wind there.  :)   But since SDL2 has historically used the older names, keeping them as a fallback makes sense if the CSS names don't work out.

```diff
diff --git a/src/video/wayland/SDL_waylandmouse.c b/src/video/wayland/SDL_waylandmouse.c
index 4abc654c2..fe6680096 100644
--- a/src/video/wayland/SDL_waylandmouse.c
+++ b/src/video/wayland/SDL_waylandmouse.c
@@ -160,6 +160,8 @@ static SDL_bool wayland_get_system_cursor(SDL_VideoData *vdata, Wayland_CursorDa
 {
     struct wl_cursor_theme *theme = NULL;
     struct wl_cursor *cursor;
+    const char *cssname = NULL;
+    const char *legacyname = NULL;
 
     char *xcursor_size;
     int size = 0;
@@ -233,46 +235,65 @@ static SDL_bool wayland_get_system_cursor(SDL_VideoData *vdata, Wayland_CursorDa
     /* Next, find the cursor from the theme... */
     switch (cdata->system_cursor) {
     case SDL_SYSTEM_CURSOR_ARROW:
-        cursor = WAYLAND_wl_cursor_theme_get_cursor(theme, "left_ptr");
+        cssname = "default";
+        legacyname = "left_ptr";
         break;
     case SDL_SYSTEM_CURSOR_IBEAM:
-        cursor = WAYLAND_wl_cursor_theme_get_cursor(theme, "xterm");
+        cssname = "text";
+        legacyname = "xterm";
         break;
     case SDL_SYSTEM_CURSOR_WAIT:
-        cursor = WAYLAND_wl_cursor_theme_get_cursor(theme, "watch");
+        cssname = "wait";
+        legacyname = "watch";
         break;
     case SDL_SYSTEM_CURSOR_CROSSHAIR:
-        cursor = WAYLAND_wl_cursor_theme_get_cursor(theme, "tcross");
+        cssname = "crosshair";
+        legacyname = "tcross";
         break;
     case SDL_SYSTEM_CURSOR_WAITARROW:
-        cursor = WAYLAND_wl_cursor_theme_get_cursor(theme, "watch");
+        cssname = "progress";
+        legacyname = "watch";
         break;
     case SDL_SYSTEM_CURSOR_SIZENWSE:
-        cursor = WAYLAND_wl_cursor_theme_get_cursor(theme, "top_left_corner");
+        cssname = "nwse-resize";
+        legacyname = "top_left_corner";
         break;
     case SDL_SYSTEM_CURSOR_SIZENESW:
-        cursor = WAYLAND_wl_cursor_theme_get_cursor(theme, "top_right_corner");
+        cssname = "nesw-resize";
+        legacyname = "top_right_corner";
         break;
     case SDL_SYSTEM_CURSOR_SIZEWE:
-        cursor = WAYLAND_wl_cursor_theme_get_cursor(theme, "sb_h_double_arrow");
+        cssname = "ew-resize";
+        legacyname = "sb_h_double_arrow";
         break;
     case SDL_SYSTEM_CURSOR_SIZENS:
-        cursor = WAYLAND_wl_cursor_theme_get_cursor(theme, "sb_v_double_arrow");
+        cssname = "ns-resize";
+        legacyname = "sb_v_double_arrow";
         break;
     case SDL_SYSTEM_CURSOR_SIZEALL:
-        cursor = WAYLAND_wl_cursor_theme_get_cursor(theme, "fleur");
+        cssname = "move";
+        legacyname = "fleur";
         break;
     case SDL_SYSTEM_CURSOR_NO:
-        cursor = WAYLAND_wl_cursor_theme_get_cursor(theme, "pirate");
+        cssname = "not-allowed";
+        legacyname = "pirate";
         break;
     case SDL_SYSTEM_CURSOR_HAND:
-        cursor = WAYLAND_wl_cursor_theme_get_cursor(theme, "hand2");
+        cssname = "pointer";
+        legacyname = "hand2";
         break;
     default:
         SDL_assert(0);
         return SDL_FALSE;
     }
 
+    cursor = WAYLAND_wl_cursor_theme_get_cursor(theme, cssname);
+
+    /* try the legacy name if the fancy new CSS name doesn't work... */
+    if (!cursor) {
+        cursor = WAYLAND_wl_cursor_theme_get_cursor(theme, legacyname);
+    }
+
     /* Fallback to the default cursor if the chosen one wasn't found */
     if (!cursor) {
         cursor = WAYLAND_wl_cursor_theme_get_cursor(theme, "left_ptr");
```

**Comment by icculus** [comment]
Also, we should probably make a `const char *GetCSSCursorName(SDL_SystemCursor c)` function in SDL_video.c that has this switch statement in one central place, since three different backends (x11, wayland, emscripten) need this mapping.

**Comment by smcv** [comment]
> Okay, I added something that _should_ fix the problem; it can load the correct cursors based on CSS theme names here, using Xcursor, but the Gnome on my system still has the old names available, so I'd appreciate it if someone could test this on a newer Gnome install, just to make sure the problem is definitely resolved.

Tested with GNOME's `adwaita-icon-theme` 46.beta on Debian unstable. It works.

One minor issue is that from source code inspection, if one of the cursors we want is missing (they're all present in Adwaita, but other themes might be less complete), the fallback is still to `left_ptr` and not `default`. Fix proposed in #9077.

> Also, we should probably make a `const char *GetCSSCursorName(SDL_SystemCursor c)` function in SDL_video.c that has this switch statement in one central place, since three different backends (x11, wayland, emscripten) need this mapping.

Done in #9077.

> EDIT: (and emscripten gets a few of these wrong in its current switch statement.)

In #9077 it uses the Wayland mapping.

**Comment by smcv** [comment]
> Note that SDL3's Wayland code already uses the CSS names exclusively with no fallback, so I'm not sure if the fallback is unnecessary for older Wayland installs or we're just throwing caution to the wind there. :) But since SDL2 has historically used the older names, keeping them as a fallback makes sense if the CSS names don't work out.

I agree that trying to be legacy-free in SDL3, but having the fallbacks in SDL2, seems proportionate.

**Comment by smcv** [comment]
> Here's an untested patch for Wayland, for SDL2.

It works. I turned it into a PR as #9078.

As with SDL3, if one of the cursors we want is missing (they're all present in Adwaita, but other themes might be less complete), the fallback is still to left_ptr and not default. Fix proposed in #9078.

I think `"move"` is actually wrong for `SIZE_ALL`: what SDL wants is an equivalent of Windows `IDC_SIZEALL`, but the CSS `move` is documented to be a drag-and-drop cursor, more like `copy` (which is the modified drag cursor that you get when you Ctrl+drag to copy a file in Windows). As an implementation detail, GNOME's Adwaita theme *does* provide a `move` that is a 4-pointed arrow, equivalent to Windows `IDC_SIZEALL`, but I think that's actually arguably a bug.

I don't see anything in the CSS spec that is a perfect representation of `IDC_SIZEALL`, so what I've suggested in #9077 and #9078 is that we use `all-scroll`, which the CSS spec at least suggests should be a 4-pointed arrow.
